In [ ]:
!apt-get update

In [8]:
!apt-get install openjdk-17-jdk-headless -qq > /dev/null

zsh:1: command not found: apt-get


In [3]:
!wget -q https://downloads.apache.org/spark/spark-4.0.1/spark-4.0.1-bin-hadoop3.tgz

zsh:1: command not found: wget


In [20]:
!tar -xvf spark-4.0.1-bin-hadoop3.tgz

spark-4.0.1-bin-hadoop3/
spark-4.0.1-bin-hadoop3/R/
spark-4.0.1-bin-hadoop3/R/lib/
spark-4.0.1-bin-hadoop3/R/lib/SparkR/
spark-4.0.1-bin-hadoop3/R/lib/SparkR/R/
spark-4.0.1-bin-hadoop3/R/lib/SparkR/R/SparkR
spark-4.0.1-bin-hadoop3/R/lib/SparkR/R/SparkR.rdb
spark-4.0.1-bin-hadoop3/R/lib/SparkR/R/SparkR.rdx
spark-4.0.1-bin-hadoop3/R/lib/SparkR/profile/
spark-4.0.1-bin-hadoop3/R/lib/SparkR/profile/general.R
spark-4.0.1-bin-hadoop3/R/lib/SparkR/profile/shell.R
spark-4.0.1-bin-hadoop3/R/lib/SparkR/tests/
spark-4.0.1-bin-hadoop3/R/lib/SparkR/tests/testthat/
spark-4.0.1-bin-hadoop3/R/lib/SparkR/tests/testthat/test_basic.R
spark-4.0.1-bin-hadoop3/R/lib/SparkR/DESCRIPTION
spark-4.0.1-bin-hadoop3/R/lib/SparkR/worker/
spark-4.0.1-bin-hadoop3/R/lib/SparkR/worker/daemon.R
spark-4.0.1-bin-hadoop3/R/lib/SparkR/worker/worker.R
spark-4.0.1-bin-hadoop3/R/lib/SparkR/doc/
spark-4.0.1-bin-hadoop3/R/lib/SparkR/doc/sparkr-vignettes.Rmd
spark-4.0.1-bin-hadoop3/R/lib/SparkR/doc/sparkr-vignettes.html
spark-4.0.

In [21]:
!pip install -q findspark

In [16]:
import os
os.environ["JAVA_HOME"] = "/usr/local/opt/openjdk@17/libexec/openjdk.jdk/Contents/Home"
os.environ["SPARK_HOME"] = "/Users/mariabelakova/Documents/учеба/projects-Data-Engineer/spark-4.0.1-bin-hadoop3"

In [17]:
import findspark
findspark.init()
from pyspark.sql import SparkSession

spark = SparkSession.builder.master("local[*]").getOrCreate()

Using Spark's default log4j profile: org/apache/spark/log4j2-defaults.properties
25/09/11 16:59:37 WARN Utils: Your hostname, MacBook-Air.local, resolves to a loopback address: 127.0.0.1; using 192.168.1.114 instead (on interface en0)
25/09/11 16:59:37 WARN Utils: Set SPARK_LOCAL_IP if you need to bind to another address
Using Spark's default log4j profile: org/apache/spark/log4j2-defaults.properties
Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
25/09/11 16:59:39 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable


In [18]:
df_iris = spark.read.csv('iris.csv', inferSchema=True, header=True)

In [19]:
df_iris.describe().show()

25/09/11 17:00:02 WARN SparkStringUtils: Truncated the string representation of a plan since it was too large. This behavior can be adjusted by setting 'spark.sql.debug.maxToStringFields'.


+-------+------------------+-------------------+------------------+------------------+---------+------------------+
|summary|      sepal_length|        sepal_width|      petal_length|       petal_width|  variety|       variety_num|
+-------+------------------+-------------------+------------------+------------------+---------+------------------+
|  count|               150|                150|               150|               150|      150|               150|
|   mean| 5.843333333333335|  3.057333333333334|3.7580000000000027| 1.199333333333334|     NULL|               1.0|
| stddev|0.8280661279778637|0.43586628493669793|1.7652982332594662|0.7622376689603467|     NULL|0.8192319205190406|
|    min|               4.3|                2.0|               1.0|               0.1|   Setosa|                 0|
|    max|               7.9|                4.4|               6.9|               2.5|Virginica|                 2|
+-------+------------------+-------------------+------------------+-----

In [20]:
from pyspark.ml.feature import StringIndexer, OneHotEncoder

In [21]:
ohe = OneHotEncoder(inputCol='variety_num', outputCol='varietyOhe')

In [22]:
model = ohe.fit(df_iris)

In [23]:
df_iris_1 = model.transform(df_iris)

In [24]:
df_iris_1.show()

+------------+-----------+------------+-----------+-------+-----------+-------------+
|sepal_length|sepal_width|petal_length|petal_width|variety|variety_num|   varietyOhe|
+------------+-----------+------------+-----------+-------+-----------+-------------+
|         5.1|        3.5|         1.4|        0.2| Setosa|          0|(2,[0],[1.0])|
|         4.9|        3.0|         1.4|        0.2| Setosa|          0|(2,[0],[1.0])|
|         4.7|        3.2|         1.3|        0.2| Setosa|          0|(2,[0],[1.0])|
|         4.6|        3.1|         1.5|        0.2| Setosa|          0|(2,[0],[1.0])|
|         5.0|        3.6|         1.4|        0.2| Setosa|          0|(2,[0],[1.0])|
|         5.4|        3.9|         1.7|        0.4| Setosa|          0|(2,[0],[1.0])|
|         4.6|        3.4|         1.4|        0.3| Setosa|          0|(2,[0],[1.0])|
|         5.0|        3.4|         1.5|        0.2| Setosa|          0|(2,[0],[1.0])|
|         4.4|        2.9|         1.4|        0.2| Se

In [25]:
from pyspark.ml.linalg import Vectors
from pyspark.ml.feature import VectorAssembler

In [26]:
df_iris_1.show()

+------------+-----------+------------+-----------+-------+-----------+-------------+
|sepal_length|sepal_width|petal_length|petal_width|variety|variety_num|   varietyOhe|
+------------+-----------+------------+-----------+-------+-----------+-------------+
|         5.1|        3.5|         1.4|        0.2| Setosa|          0|(2,[0],[1.0])|
|         4.9|        3.0|         1.4|        0.2| Setosa|          0|(2,[0],[1.0])|
|         4.7|        3.2|         1.3|        0.2| Setosa|          0|(2,[0],[1.0])|
|         4.6|        3.1|         1.5|        0.2| Setosa|          0|(2,[0],[1.0])|
|         5.0|        3.6|         1.4|        0.2| Setosa|          0|(2,[0],[1.0])|
|         5.4|        3.9|         1.7|        0.4| Setosa|          0|(2,[0],[1.0])|
|         4.6|        3.4|         1.4|        0.3| Setosa|          0|(2,[0],[1.0])|
|         5.0|        3.4|         1.5|        0.2| Setosa|          0|(2,[0],[1.0])|
|         4.4|        2.9|         1.4|        0.2| Se

In [27]:
df_iris_1.columns

['sepal_length',
 'sepal_width',
 'petal_length',
 'petal_width',
 'variety',
 'variety_num',
 'varietyOhe']

In [28]:
assembler = VectorAssembler(inputCols=['sepal_length', 'sepal_width', 'petal_length', 'petal_width'], outputCol='Features')

In [29]:
df_iris_1 = assembler.transform(df_iris_1)

In [30]:
df_iris_1.show()

+------------+-----------+------------+-----------+-------+-----------+-------------+-----------------+
|sepal_length|sepal_width|petal_length|petal_width|variety|variety_num|   varietyOhe|         Features|
+------------+-----------+------------+-----------+-------+-----------+-------------+-----------------+
|         5.1|        3.5|         1.4|        0.2| Setosa|          0|(2,[0],[1.0])|[5.1,3.5,1.4,0.2]|
|         4.9|        3.0|         1.4|        0.2| Setosa|          0|(2,[0],[1.0])|[4.9,3.0,1.4,0.2]|
|         4.7|        3.2|         1.3|        0.2| Setosa|          0|(2,[0],[1.0])|[4.7,3.2,1.3,0.2]|
|         4.6|        3.1|         1.5|        0.2| Setosa|          0|(2,[0],[1.0])|[4.6,3.1,1.5,0.2]|
|         5.0|        3.6|         1.4|        0.2| Setosa|          0|(2,[0],[1.0])|[5.0,3.6,1.4,0.2]|
|         5.4|        3.9|         1.7|        0.4| Setosa|          0|(2,[0],[1.0])|[5.4,3.9,1.7,0.4]|
|         4.6|        3.4|         1.4|        0.3| Setosa|     

Или получим все колонки др способом через Pipeline



In [31]:
from pyspark.ml import Pipeline

In [32]:
pipeline = Pipeline(stages =
[
  OneHotEncoder(inputCol='variety_num', outputCol='varietyOhe'),
  VectorAssembler(inputCols=['sepal_length', 'sepal_width', 'petal_length', 'petal_width', 'variety_num', 'varietyOhe'], outputCol='Features')
]
)

In [33]:
pipelineTrained = pipeline.fit(df_iris)

In [34]:
pipelineTrained.transform(df_iris).show()

+------------+-----------+------------+-----------+-------+-----------+-------------+--------------------+
|sepal_length|sepal_width|petal_length|petal_width|variety|variety_num|   varietyOhe|            Features|
+------------+-----------+------------+-----------+-------+-----------+-------------+--------------------+
|         5.1|        3.5|         1.4|        0.2| Setosa|          0|(2,[0],[1.0])|[5.1,3.5,1.4,0.2,...|
|         4.9|        3.0|         1.4|        0.2| Setosa|          0|(2,[0],[1.0])|[4.9,3.0,1.4,0.2,...|
|         4.7|        3.2|         1.3|        0.2| Setosa|          0|(2,[0],[1.0])|[4.7,3.2,1.3,0.2,...|
|         4.6|        3.1|         1.5|        0.2| Setosa|          0|(2,[0],[1.0])|[4.6,3.1,1.5,0.2,...|
|         5.0|        3.6|         1.4|        0.2| Setosa|          0|(2,[0],[1.0])|[5.0,3.6,1.4,0.2,...|
|         5.4|        3.9|         1.7|        0.4| Setosa|          0|(2,[0],[1.0])|[5.4,3.9,1.7,0.4,...|
|         4.6|        3.4|         1.

In [35]:
df_iris_2 = pipelineTrained.transform(df_iris)

In [36]:
df_iris_2.show()

+------------+-----------+------------+-----------+-------+-----------+-------------+--------------------+
|sepal_length|sepal_width|petal_length|petal_width|variety|variety_num|   varietyOhe|            Features|
+------------+-----------+------------+-----------+-------+-----------+-------------+--------------------+
|         5.1|        3.5|         1.4|        0.2| Setosa|          0|(2,[0],[1.0])|[5.1,3.5,1.4,0.2,...|
|         4.9|        3.0|         1.4|        0.2| Setosa|          0|(2,[0],[1.0])|[4.9,3.0,1.4,0.2,...|
|         4.7|        3.2|         1.3|        0.2| Setosa|          0|(2,[0],[1.0])|[4.7,3.2,1.3,0.2,...|
|         4.6|        3.1|         1.5|        0.2| Setosa|          0|(2,[0],[1.0])|[4.6,3.1,1.5,0.2,...|
|         5.0|        3.6|         1.4|        0.2| Setosa|          0|(2,[0],[1.0])|[5.0,3.6,1.4,0.2,...|
|         5.4|        3.9|         1.7|        0.4| Setosa|          0|(2,[0],[1.0])|[5.4,3.9,1.7,0.4,...|
|         4.6|        3.4|         1.

In [37]:
df_iris.show()

+------------+-----------+------------+-----------+-------+-----------+
|sepal_length|sepal_width|petal_length|petal_width|variety|variety_num|
+------------+-----------+------------+-----------+-------+-----------+
|         5.1|        3.5|         1.4|        0.2| Setosa|          0|
|         4.9|        3.0|         1.4|        0.2| Setosa|          0|
|         4.7|        3.2|         1.3|        0.2| Setosa|          0|
|         4.6|        3.1|         1.5|        0.2| Setosa|          0|
|         5.0|        3.6|         1.4|        0.2| Setosa|          0|
|         5.4|        3.9|         1.7|        0.4| Setosa|          0|
|         4.6|        3.4|         1.4|        0.3| Setosa|          0|
|         5.0|        3.4|         1.5|        0.2| Setosa|          0|
|         4.4|        2.9|         1.4|        0.2| Setosa|          0|
|         4.9|        3.1|         1.5|        0.1| Setosa|          0|
|         5.4|        3.7|         1.5|        0.2| Setosa|     

In [38]:
train, test = df_iris_1.randomSplit([0.8, 0.2], seed=12345)

In [39]:
train.show()

+------------+-----------+------------+-----------+----------+-----------+-------------+-----------------+
|sepal_length|sepal_width|petal_length|petal_width|   variety|variety_num|   varietyOhe|         Features|
+------------+-----------+------------+-----------+----------+-----------+-------------+-----------------+
|         4.3|        3.0|         1.1|        0.1|    Setosa|          0|(2,[0],[1.0])|[4.3,3.0,1.1,0.1]|
|         4.4|        2.9|         1.4|        0.2|    Setosa|          0|(2,[0],[1.0])|[4.4,2.9,1.4,0.2]|
|         4.4|        3.0|         1.3|        0.2|    Setosa|          0|(2,[0],[1.0])|[4.4,3.0,1.3,0.2]|
|         4.4|        3.2|         1.3|        0.2|    Setosa|          0|(2,[0],[1.0])|[4.4,3.2,1.3,0.2]|
|         4.5|        2.3|         1.3|        0.3|    Setosa|          0|(2,[0],[1.0])|[4.5,2.3,1.3,0.3]|
|         4.6|        3.1|         1.5|        0.2|    Setosa|          0|(2,[0],[1.0])|[4.6,3.1,1.5,0.2]|
|         4.6|        3.4|         1.

## Создадим модель логистической регреcсии и обучим её.

In [40]:
from pyspark.ml.classification import LogisticRegression

In [ ]:
lr = LogisticRegression(featuresCol = 'Features', labelCol = 'variety_num', maxIter=100)
lrModel = lr.fit(train)

25/09/11 17:01:08 WARN InstanceBuilder: Failed to load implementation from:dev.ludovic.netlib.blas.JNIBLAS


In [42]:
train_res = lrModel.transform(train)
test_res = lrModel.transform(test)

In [43]:
train_res.show()

+------------+-----------+------------+-----------+----------+-----------+-------------+-----------------+--------------------+--------------------+----------+
|sepal_length|sepal_width|petal_length|petal_width|   variety|variety_num|   varietyOhe|         Features|       rawPrediction|         probability|prediction|
+------------+-----------+------------+-----------+----------+-----------+-------------+-----------------+--------------------+--------------------+----------+
|         4.3|        3.0|         1.1|        0.1|    Setosa|          0|(2,[0],[1.0])|[4.3,3.0,1.1,0.1]|[67.7027916872411...|[1.0,1.6424725044...|       0.0|
|         4.4|        2.9|         1.4|        0.2|    Setosa|          0|(2,[0],[1.0])|[4.4,2.9,1.4,0.2]|[57.0087432508377...|[1.0,1.4036232460...|       0.0|
|         4.4|        3.0|         1.3|        0.2|    Setosa|          0|(2,[0],[1.0])|[4.4,3.0,1.3,0.2]|[61.7475607768861...|[1.0,2.4772188981...|       0.0|
|         4.4|        3.2|         1.3| 

Воспользуемся MulticlassClassificationEvaluator для оценки качества на train и test множестве.

In [44]:
from pyspark.ml.evaluation import MulticlassClassificationEvaluator

In [48]:
ev = MulticlassClassificationEvaluator(labelCol='variety_num', predictionCol='prediction', metricName='accuracy')

In [51]:
accuracy_train = ev.evaluate(train_res)
accuracy_test = ev.evaluate(test_res)

print(f"Train accuracy: {accuracy_train:.3f}")
print(f"Test accuracy: {accuracy_test:.3f}")

Train accuracy: 0.984
Test accuracy: 1.000


## Выводы

- Модель хорошо обучена
- Очень высокая точность на тренировке показывает, что признаки (длина и ширина чашелистика и лепестка) позволяют эффективно различать классы ирисов.
- Нет признаков переобучения
- Если бы точность на тренировке была сильно выше, чем на тесте, это сигнал переобучения. Здесь тестовая точность даже 100%, значит модель хорошо обобщает.
- Iris dataset небольшой (150 записей, 3 класса) и линейно разделимый. Поэтому логистическая регрессия показывает почти идеальные результаты.
- Модель подходит для данной задачи
- Можно использовать такую модель для классификации ирисов на основе этих 4 признаков без дополнительной настройки.

# Итог
Модель логистической регрессии на признаках длины и ширины чашелистика и лепестка очень эффективно классифицирует ирисы. Переобучения нет, модель хорошо обобщает на тестовой выборке.